cli/cmd/identity.go

```diff
@@ -0,0 +1,295 @@
+
+					fmt.Printf("\n%s", resultCerticate)
+				} else {
+					fmt.Printf("\n%s\n", result.err)
+				}
+			}
+			return nil
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace to use for --proxy versions (default: all namespaces)")
+	cmd.PersistentFlags().StringVarP(&options.selector, "selector", "l", options.selector, "Selector (label query) to filter on, supports ‘=’, ‘==’, and ‘!=’ ")
+	return cmd
+}
+
+// getCertificate fetches the certificates for all the pods
+func getCertificate(k8sAPI *k8s.KubernetesAPI, pods []corev1.Pod, portName string, waitingTime time.Duration, emitLog bool) []certificate {
+	var certificates []certificate
+	resultChan := make(chan certificate)
+	var activeRoutines int32
+	for _, pod := range pods {
+		atomic.AddInt32(&activeRoutines, 1)
+		go func(p corev1.Pod) {
+			defer atomic.AddInt32(&activeRoutines, -1)
+			containers, err := getContainersWithPort(p, portName)
+			if err != nil {
+				resultChan <- certificate{
+					pod: p.GetName(),
+					err: err,
+				}
+				return
+			}
+
+			for _, c := range containers {
+				cert, err := getContainerCertificate(k8sAPI, p, c, portName, emitLog)
+				resultChan <- certificate{
+					pod:         p.GetName(),
+					container:   c.Name,
+					Certificate: cert,
+					err:         err,
+				}
+			}
+		}(pod)
+	}
+
+	for {
+		select {
+		case cert := <-resultChan:
+			certificates = append(certificates, cert)
+		case <-time.After(waitingTime):
+			break
+		}
+		if atomic.LoadInt32(&activeRoutines) == 0 {
+			break
+		}
+	}
+	return certificates
+}
+
+func getContainersWithPort(pod corev1.Pod, portName string) ([]corev1.Container, error) {
+	if pod.Status.Phase != corev1.PodRunning {
+		return nil, fmt.Errorf("pod not running: %s", pod.GetName())
+	}
+	var containers []corev1.Container
+
+	for _, c := range pod.Spec.Containers {
+		for _, p := range c.Ports {
+			if p.Name == portName {
+				containers = append(containers, c)
+			}
+		}
+	}
+	return containers, nil
+}
+
+func getContainerCertificate(k8sAPI *k8s.KubernetesAPI, pod corev1.Pod, container corev1.Container, portName string, emitLog bool) ([]*x509.Certificate, error) {
+	portForward, err := k8s.NewContainerMetricsForward(k8sAPI, pod, container, emitLog, portName)
+	if err != nil {
+		return nil, err
+	}
+
+	defer portForward.Stop()
+	if err = portForward.Init(); err != nil {
+		fmt.Fprintf(os.Stderr, "Error running port-forward: %s", err)
+		return nil, err
+	}
+
+	certURL := portForward.URLFor("")
+	return getCertResponse(certURL, pod)
+}
+
+func getCertResponse(url string, pod corev1.Pod) ([]*x509.Certificate, error) {
+	serverName, err := getServerName(pod, "linkerd-proxy")
+	if err != nil {
+		return nil, err
+	}
+	rootPEM, err := getTrustAnchor(pod, "linkerd-proxy")
+	if err != nil {
+		return nil, err
+	}
+
+	re := regexp.MustCompile("[0-9]+")
+	res := re.FindString(url)
+
+	connURL := "localhost:" + res
```

Comment:
Can you explain what's going on here?  We're extracting the port from the port-forward url and appending it onto localhost?  Isn't the port-forward url already a localhost address?

Reply:
The thing is that the `tls.Dial` function in Go expects a domain name with port and not the entire url, and `portForward.URLFor()` gives us the URL. Just had to trim the `http://` part from the url, which I now did using the `strings.Trim`. I agree this was an overkill and changed it.